### PR TITLE
PAINTROID-304 Text field size is reset if font or font-effect (bold, italic, ...) is changed

### PR DESCRIPTION
**Pull request by Electronix1337**
After this ticket the texttoolbox will not reset when font or font-effect is changed. Had to change some tests in the TextToolIntegrationTest suite. 

Additionally I did find a bug/unexpected behaviour where the text will be displayed slightly outside of the text tool box which I also tried to fix in this ticket but not sure if the result is perfect. My approach was to adjust the boxwidth when italic is applied.

https://jira.catrob.at/browse/PAINTROID-304

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Paintroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [x] Post a message in the *#paintroid* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Review by JulianJautz — DISMISSED** [review]
As far as I understand the description in the ticket, the goal is that the size of the textbox doesn't change after the font was set to bold, italic,... (and not the font size itself). This is not what your implementation does.
Do I miss something?

And I also noticed that this change in size also happens, when the font is changed or "text" in the bottom bar is clicked, so always when something in the text tool related menu is changed.

**Comment by Electronix1337** [comment]
> As far as I understand the description in the ticket, the goal is that the size of the textbox doesn't change after the font was set to bold, italic,... (and not the font size itself). This is not what your implementation does. Do I miss something?
> 
> And I also noticed that this change in size also happens, when the font is changed or "text" in the bottom bar is clicked, so always when something in the text tool related menu is changed.

You are right, but I talked to Thorsten before I started implementing the ticket and we both came to the same conclusion that the only reasonable solution to this would be to automatically adjust the font size when text box is resized using corner points. As the changes in font type and text style are applied to the text with the internal font size I would have to calculate the proportional difference when resizing but this seems like an overkill to me or rather counterintuitive in comparison to my implementation. I also don't see a proper use case where users want to distort the text field only in X or Y (so all resizing points except the corners) and then change the font type or text style as a distorted text is rarely the desired outcome. So in short, if a user resizes the text box and applies a change in font type or text style the desired outcome would be that the text box stays the same in size, which happens with my implementation and for this to properly work, I need to internally adjust the font size as the changes are made based on this. Hope it's clear what I mean with this.

**Comment by JulianJautz** [comment]
I think I understand. So resizing via the corners seems to work as you described.
And the text box shrinkage after setting bold, italic,... won't be changed (when resizing box before only in x or y direction)?

> > And I also noticed that this change in size also happens, when the font is changed or "text" in the bottom bar is clicked, so always when something in the text tool related menu is changed.
> 
> I don't really get what you mean with this, could you further describe that please.
I meant that the box also shrinks when clicking on "Tools" or "Text" in the footer (box size change in x or y direction before). I think it would be good to not change the text box size when clicking these buttons.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt [hunk 1]
```diff
@@ -97,9 +99,14 @@ class TextTool(
...
 
-    private var textSize = DEFAULT_TEXT_SIZE
+    var textSize = DEFAULT_TEXT_SIZE
     private val stc: Typeface?
     private val dubai: Typeface?
+    var callback: TextToolOptionsView.Callback? = null
+    var oldBoxWidth = 0f
+    var oldBoxHeight = 0f
+    var oldToolPosition: PointF? = null
+    var wasItalicRecentlyDisabled = false
```
Comment (StanglTobias) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt:109]: can be removed

**Review by StanglTobias — CHANGES_REQUESTED** [review]
There are still some problems when the text is italic

Screenshot_20220622_171909

**Review by StanglTobias — APPROVED** [review]

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt [hunk 2]
```diff
@@ -198,15 +225,25 @@ class TextTool(
...
+        } else {
+            multilineText.forEachIndexed { index, textLine ->
+                canvas.drawText(
+                    textLine,
+                    -(scaledBoxWidth / 2) + scaledWidthOffset,
+                    -(scaledBoxHeight / 2) + scaledHeightOffset - textAscent + lineHeight * index,
+                    textPaint
+                )
+            }
         }
```
Comment (CodeChamp-SS) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt:246]: You can write this also as suggested in `TextToolCommand.kt`.

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt [hunk 3]
```diff
@@ -218,11 +255,43 @@ class TextTool(
...
+
+    private fun applyAttributes() {
+        boxWidth = oldBoxWidth
+        boxHeight = oldBoxHeight
+        if (oldToolPosition != null) {
+            toolPosition = oldToolPosition as PointF
+        } else {
+            resetBoxPosition()
+        }
+    }
```
Comment (CodeChamp-SS) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt:293]:
```suggestion
    private fun storeAttributes(italic: Boolean = false) {
        if (italic) {
            boxWidth *= ITALIC_FONT_BOX_ADJUSTMENT
        }
        oldBoxWidth = boxWidth
        oldBoxHeight = boxHeight
        oldToolPosition = PointF(toolPosition.x, toolPosition.y)
    }

    private fun applyAttributes(italic: Boolean = false) {
        boxWidth = oldBoxWidth / if (italic) ITALIC_FONT_BOX_ADJUSTMENT else 1f
        boxHeight = oldBoxHeight
        if (oldToolPosition != null) {
            toolPosition = oldToolPosition as PointF
        } else {
            resetBoxPosition()
        }
    }
```
You can write it like this, to remove duplicity and clean up the code.

**Review by CodeChamp-SS — CHANGES_REQUESTED** [review]
The changes work as expected. Please look into the suggested changes (just for cleaning up the code a bit). Then I'll approve the PR.

File: Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/TextToolCommand.kt [hunk 4]
```diff
@@ -69,13 +74,24 @@ class TextToolCommand(
...
+                }
+            } else {
+                multilineText.forEachIndexed { index, textLine ->
+                    canvas.drawText(
+                        textLine,
+                        -(scaledBoxWidth / 2) + scaledWidthOffset,
+                        -(scaledBoxHeight / 2) + scaledHeightOffset - textAscent + lineHeight * index,
+                        textPaint
+                    )
+                }
```
[the remark this answers is not in the document]
Comment (CodeChamp-SS) [reply Paintroid/src/main/java/org/catrobat/paintroid/command/implementation/TextToolCommand.kt:94]: I think you also remove line 77 and use that directly inside the `drawText()` method in place of that `x`.

**Review by CodeChamp-SS — APPROVED** [review]
LGTM!

File: Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt [hunk 5]
```diff
@@ -100,9 +102,13 @@ class TextTool(
     @JvmField
     var bold = false
 
-    private var textSize = DEFAULT_TEXT_SIZE
+    var textSize = DEFAULT_TEXT_SIZE
     private val stc: Typeface?
     private val dubai: Typeface?
+    var callback: TextToolOptionsView.Callback? = null
```
Comment (CodeChamp-SS) [inline Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/TextTool.kt:108]: I think this `callback` var can be removed, as it isn't getting used anywhere.

**Review by CodeChamp-SS — COMMENTED** [review]
Please look into these other minor changes :)

**Review by CodeChamp-SS — APPROVED** [review]

**Pull request merged**